### PR TITLE
feat: Pebble Notices

### PR DESCRIPTION
**Pull request by kayra1**
# Description

This change introduces pebble notices to the GoCert's backend as included in spec TE034. It is enabled in the configuration file. Each update to a CSR's related certificate will create a new pebble notice, which will allow the charm to listen to changes in GoCert and make relevant changes to the charm relations.

The test for this feature is inside of build_rock.yaml. This is both because testing the execution of external commands is difficult without mocking, and to make sure that there is no doubt that pebble is actually receiving the request in a regular rock deployment.

`$ docker exec gocert /usr/bin/pebble notices`
```
ID   User    Type           Key  First               Repeated            Occurrences
1    public  change-update  1    today at 07:32 UTC  today at 07:32 UTC  4
2    public  change-update  2    today at 07:32 UTC  today at 07:32 UTC  3
```

`$ <submit CSR through curl>`
**stdout**:
```
% Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0     0 --:--:-- --:--:-- --:--:--     0
100  1082  100     1  100  1081    147  155k --:--:-- --:--:-- --:--:--  176k
```
**stderr**: `1`

`$ <submit certificate through curl>`

**stdout**:
```
% Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  1334  100     1  100  1333     99   129k --:--:-- --:--:-- --:--:--  144k
```

**stderr**: `1`

`$ docker exec gocert /usr/bin/pebble notices`
```
1ID   User    Type           Key                            First               Repeated            Occurrences
1    public  change-update  1                              today at 07:32 UTC  today at 07:32 UTC  4
2    public  change-update  2                              today at 07:32 UTC  today at 07:32 UTC  3
3    0       custom         gocert.com/certificate/create  today at 07:32 UTC  today at 07:32 UTC  1
```

When the pebble binary is missing or any other error is produced, an error is logged, but if the change was successfully committed to the database, the GoCert application does not roll back the changes and responds to the HTTP request regularly. This is to support deployments outside of the Rock or any custom containers without pebble installed. It's also not preferable for users to go through the effort of uploading a cert, get everything right, then get rejected for something that's out of their control.

# Checklist:

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that validate the behaviour of the software
- [x] I validated that new and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by kayra1** [comment]
It seems that pebble notices don't work without the primary path of the key not ending with a TLD, so I've changed it to be gocert.com. I'm also not fully convinced about the key being called `create`, I think there is a lot of use from having the key being called `update` and the notice being fired when a certificate is replaced or deleted or rejected after it was created.

File: README.md [hunk 1]
```diff
@@ -15,11 +15,25 @@ GoCert requires 3 files to operate:
...
+
+### Config File
+The config file requires the following parameters:
+* keypath: this is the path to the private key you've generated.
+* certpath: this is the path to a certificate for enabling HTTPS connections.
+* dbpath: the path to a sqlite database file. If the file does not exist GoCert will attempt to create it.
+* port: the port in integer form to serve all of GoCert's API and frontend
+* pebblenotificationsenabled: a boolean that once enabled, will allow GoCert to send pebble notices. Read more about it (here)[https://github.com/canonical/pebble?tab=readme-ov-file#notices].
+
+an example config file may look like:
```
Comment (ghislainbourgeois) [inline README.md:31]:
```suggestion
An example config file may look like:
```

File: README.md [hunk 2]
```diff
@@ -15,11 +15,25 @@ GoCert requires 3 files to operate:
 * A private key
 * A TLS certificate with that private key
 * A YAML config file with the required parameters
-as an example:
+
+you can generate the cert and the associated key by running:
```
Comment (ghislainbourgeois) [inline README.md:19]:
```suggestion
You can generate the cert and the associated key by running:
```

**Review by ghislainbourgeois — CHANGES_REQUESTED** [review]
2 small nitpicks, otherwise LGTM

**Review by ghislainbourgeois — APPROVED** [review]

File: internal/api/server.go [hunk 3]
```diff
@@ -65,9 +69,18 @@ func validateConfigFile(filePath string) (Config, error) {
...
 }
 
+func SendPebbleNotification(key, request_id string) error {
+		cmd := exec.Command("pebble", "notify", key, fmt.Sprintf("request_id=%s", request_id))
+		if err := cmd.Run(); err != nil {
+			return errors.Join(errors.New("couldn't execute a pebble notify: "), err)
+		}
+		return nil
+}
+
```
Comment (gruyaume) [inline internal/api/server.go:83]: Why so much indentation?

If you use vscode you can use auto format on save to avoid this

File: README.md [hunk 4]
```diff
@@ -15,11 +15,25 @@ GoCert requires 3 files to operate:
...
+`openssl req -newkey rsa:2048 -nodes -keyout key.pem -x509 -days 1 -out cert.pem -subj "/CN=example.com"`
+GoCert does not support insecure http connections.
+
+### Config File
+The config file requires the following parameters:
+* keypath: this is the path to the private key you've generated.
+* certpath: this is the path to a certificate for enabling HTTPS connections.
+* dbpath: the path to a sqlite database file. If the file does not exist GoCert will attempt to create it.
+* port: the port in integer form to serve all of GoCert's API and frontend
+* pebblenotificationsenabled: a boolean that once enabled, will allow GoCert to send pebble notices. Read more about it (here)[https://github.com/canonical/pebble?tab=readme-ov-file#notices].
```
Comment (gruyaume) [inline README.md:29]: Adding underscores would improve readability. I'd recommend `pebble_notify` instead of `pebblenotificationsenabled`. The same applies to other parameters (ex. `key_path` instead of `keypath`), but those are likely not part of this PR's scope

File: README.md [hunk 5]
```diff
@@ -15,11 +15,25 @@ GoCert requires 3 files to operate:
...
+You can generate the cert and the associated key by running:
+`openssl req -newkey rsa:2048 -nodes -keyout key.pem -x509 -days 1 -out cert.pem -subj "/CN=example.com"`
+GoCert does not support insecure http connections.
+
+### Config File
+The config file requires the following parameters:
+* keypath: this is the path to the private key you've generated.
+* certpath: this is the path to a certificate for enabling HTTPS connections.
+* dbpath: the path to a sqlite database file. If the file does not exist GoCert will attempt to create it.
+* port: the port in integer form to serve all of GoCert's API and frontend
```
Comment (gruyaume) [inline README.md:28]:
```suggestion
* port: port number on which GoCert will listen for all incoming API and frontend connections (integer).
```

File: README.md [hunk 6]
```diff
@@ -15,11 +15,25 @@ GoCert requires 3 files to operate:
...
+
+You can generate the cert and the associated key by running:
+`openssl req -newkey rsa:2048 -nodes -keyout key.pem -x509 -days 1 -out cert.pem -subj "/CN=example.com"`
+GoCert does not support insecure http connections.
+
+### Config File
+The config file requires the following parameters:
+* keypath: this is the path to the private key you've generated.
+* certpath: this is the path to a certificate for enabling HTTPS connections.
+* dbpath: the path to a sqlite database file. If the file does not exist GoCert will attempt to create it.
```
Comment (gruyaume) [inline README.md:27]:
```suggestion
* dbpath: path to a sqlite database file. If the file does not exist GoCert will attempt to create it. (string)
```

File: README.md [hunk 7]
```diff
@@ -15,11 +15,25 @@ GoCert requires 3 files to operate:
...
-as an example:
+
+You can generate the cert and the associated key by running:
+`openssl req -newkey rsa:2048 -nodes -keyout key.pem -x509 -days 1 -out cert.pem -subj "/CN=example.com"`
+GoCert does not support insecure http connections.
+
+### Config File
+The config file requires the following parameters:
+* keypath: this is the path to the private key you've generated.
+* certpath: this is the path to a certificate for enabling HTTPS connections.
```
Comment (gruyaume) [inline README.md:26]:
```suggestion
* certpath: path to a PEM formatted certificate for enabling HTTPS connections (string).
```

File: README.md [hunk 8]
```diff
@@ -15,11 +15,25 @@ GoCert requires 3 files to operate:
...
 * A YAML config file with the required parameters
-as an example:
+
+You can generate the cert and the associated key by running:
+`openssl req -newkey rsa:2048 -nodes -keyout key.pem -x509 -days 1 -out cert.pem -subj "/CN=example.com"`
+GoCert does not support insecure http connections.
+
+### Config File
+The config file requires the following parameters:
+* keypath: this is the path to the private key you've generated.
```
Comment (gruyaume) [inline README.md:25]:
```suggestion
* keypath: path to the private key for enabling HTTPS connections (string).
```

File: README.md (hunk 4 above)
Comment (gruyaume) [inline README.md:29]:
```suggestion
* pebble_notify: Allow GoCert to send pebble notices on certificate events (create, update, delete). Read more about it (here)[https://github.com/canonical/pebble?tab=readme-ov-file#notices]. (bool)
```

File: internal/api/server.go (hunk 3 above)
Comment (kayra1) [reply internal/api/server.go:83]: i think i forgot to enable it when I moved computers. fixed.

File: README.md (hunk 4 above)
Comment (kayra1) [reply README.md:29]: changed all the params here, since it's pretty easy to remap them. The config stuff can use some cleaning up still, I've opened another ticket for that. TLSENG-270

Comment (kayra1) [reply README.md]: Took all of the suggestions and converted this into a table. [Here's](https://github.com/canonical/gocert/blob/TLSENG-239-pebble-notices/README.md) the preview.

File: README.md (hunk 8 above)
Comment (kayra1) [reply README.md]: Took all of the suggestions and converted this into a table. [Here's](https://github.com/canonical/gocert/blob/TLSENG-239-pebble-notices/README.md) the preview.

File: README.md (hunk 7 above)
Comment (kayra1) [reply README.md]: Took all of the suggestions and converted this into a table. [Here's](https://github.com/canonical/gocert/blob/TLSENG-239-pebble-notices/README.md) the preview.

File: README.md (hunk 6 above)
Comment (kayra1) [reply README.md]: Took all of the suggestions and converted this into a table. [Here's](https://github.com/canonical/gocert/blob/TLSENG-239-pebble-notices/README.md) the preview.

File: internal/api/handlers.go
Comment (saltiyazan) [inline internal/api/handlers.go:1]: Why do we send pebble notifications in changes on certificates and we don't do the same for CSRs?

**Comment by saltiyazan** [comment]
> When the pebble binary is missing or any other error is produced, an error is logged, but if the change was successfully committed to the database, the GoCert application does not roll back the changes and responds to the HTTP request regularly. This is to support deployments outside of the Rock or any custom containers without pebble installed. It's also not preferable for users to go through the effort of uploading a cert, get everything right, then get rejected for something that's out of their control.

What if there was an error and we couldn't send a pebble notification for this specific certificate to the charm? would the charm still be able to get it later somehow? or will the charm miss that change?

**Comment by kayra1** [comment]
> > When the pebble binary is missing or any other error is produced, an error is logged, but if the change was successfully committed to the database, the GoCert application does not roll back the changes and responds to the HTTP request regularly. This is to support deployments outside of the Rock or any custom containers without pebble installed. It's also not preferable for users to go through the effort of uploading a cert, get everything right, then get rejected for something that's out of their control.
> 
> What if there was an error and we couldn't send a pebble notification for this specific certificate to the charm? would the charm still be able to get it later somehow? or will the charm miss that change?

Any error that may occur with pebble is simply out of our control, we provide a rock with pebble guaranteed to be installed. So if any error were to occur, the cause would be pebble/charm specific or container image specific.

If you mean intermittent runtime errors, pebble notifications not working would mean pebble is broken. This would break a lot of things with charms, like replans and restarts or any container interactions.

In this case GoCert notices are just as reliable as the infrastructure that supports it, and there's not much we could do about that.

So yes, long story short charms will miss that event.

File: internal/api/handlers.go
Comment (kayra1) [reply internal/api/handlers.go:1]: That's how it is in the spec + we expect the CSR to come from the charm relation interface, so notifying it back doesn't make too much sense at the moment. It could make sense if a CSR was manually added, but that means there is no relation to provide its certificate to.

File: README.md [hunk 9]
```diff
@@ -15,11 +15,30 @@ GoCert requires 3 files to operate:
...
+GoCert does not support insecure http connections.
+
+### Config File
+The config file requires the following parameters:
+| Key                  | Type    | Description |
+|----------------------|---------|----------|
+| key_path             | string  | path to the private key for enabling HTTPS connections |
+| cert_path            | string  | path to a PEM formatted certificate for enabling HTTPS connections |
+| db_path              | string  | path to a sqlite database file. If the file does not exist GoCert will attempt to create it. |
+| port                 | number  | port number on which GoCert will listen for all incoming API and frontend connections. |
```
Comment (gruyaume) [inline README.md:33]:
```suggestion
| port                 | integer  | port number on which GoCert will listen for all incoming API and frontend connections. |
```

File: README.md [hunk 10]
```diff
@@ -15,11 +15,30 @@ GoCert requires 3 files to operate:
...
+| pebble_notifications | boolean | Allow GoCert to send pebble notices on certificate events (create, update, delete). Read more about it [here](https://github.com/canonical/pebble?tab=readme-ov-file#notices). |
```
Comment (gruyaume) [inline README.md:34]:
```suggestion
| pebble_notifications | boolean | Allow GoCert to send pebble notices on certificate events (create, update, delete). Pebble needs to be running on the same system as GoCert. Read more about Pebble Notices [here](https://github.com/canonical/pebble?tab=readme-ov-file#notices). |
```

**Review by gruyaume — APPROVED** [review]

File: internal/api/handlers.go
Comment (saltiyazan) [reply internal/api/handlers.go:1]: Yeah, if only the charm will do things with CSRs then sure. Thanks.

**Review by saltiyazan — APPROVED** [review]
